libfoundation/src/system-library-android.hpp

```diff
@@ -0,0 +1,168 @@
+/*                                                                     -*-c++-*-
+ Copyright (C) 2017 LiveCode Ltd.
+ 
+ This file is part of LiveCode.
+ 
+ LiveCode is free software; you can redistribute it and/or modify it under
+ the terms of the GNU General Public License v3 as published by the Free
+ Software Foundation.
+ 
+ LiveCode is distributed in the hope that it will be useful, but WITHOUT ANY
+ WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+
+#include "system-private.h"
+
+#include <foundation-auto.h>
+
+/* ================================================================
+ * Android Handle Class
+ * ================================================================ */
+
+#include "system-library-posix.hpp"
+
+#include <dlfcn.h>
+#include <link.h>
+
+#if __ANDROID_API__ < 21
+extern "C" __attribute__((weak)) int
+dl_iterate_phdr(int (*)(struct dl_phdr_info *, size_t, void *),
+                void *);
+#endif
+
+extern MCStringRef gMCSAndroidNativeLibPath;
+
+class __MCSLibraryHandleAndroid: public __MCSLibraryHandlePosix
+{
+public:
+    bool
+    CopyNativePath(MCStringRef& r_native_path) const
+    {
+        const char *t_sys_filename = nullptr;
+        if (dl_iterate_phdr == nullptr)
+        {
+            t_sys_filename = CopyNativePathLegacy();
+        }
+        else
+        {
+            t_sys_filename = CopyNativePathModern(m_handle).Run();
+        }
+
+        if (t_sys_filename != nullptr)
+        {
+            
+            MCAutoStringRef t_filename;
+            if (!MCStringCreateWithSysString(t_sys_filename,
+                                             &t_filename))
+            {
+                return false;
+            }
+        
+            return MCStringFormat(r_native_path,
+                                  "%@/%@",
+                                  gMCSAndroidNativeLibPath,
+                                  *t_filename);
+        }
+
+        return __MCSLibraryThrowResolvePathFailed();
+    }
+    
+private:
+    class CopyNativePathModern
+    {
+    public:
+        CopyNativePathModern(void *p_handle)
+            : m_handle(p_handle),
+              m_filename(nullptr)
+        {
+        }
+        
+        const char *Run(void)
+        {
+            dl_iterate_phdr(Thunk,
+                            this);
+            return m_filename;
+        }
```

Comment:
Consider using a lambda here rather than requiring a `Thunk()` static method.

```c++
const char *Run()
{
    dl_iterate_phdr([](struct dl_phdr_info* p_info, size_t p_size, void *p_context) {
                        auto t_func = static_cast<CopyNativePathModern *>(p_context);
                        return (*t_func)(p_info, p_size);
                    }, this);
    return m_filename;
}
```